### PR TITLE
plan: new plan supports join.

### PR DESCRIPTION
**Pull request by hanfei1991**
This PR let new plan support join. We only introduce hash join now, and in order to simplify problem, hash join will block any physical property.
@shenli @coocood @winoros @lamxTyler PTAL

### CODE REVIEW

File: plan/new_physical_plan_builder.go [hunk 1]
```diff
@@ -98,6 +98,91 @@ func (p *Projection) convert2NewPhysicalPlan(prop *requiredProp) (taskProfile, e
...
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	switch p.JoinType {
+	case LeftOuterJoin:
+		hashJoin.SmallTable = 1
+	case RightOuterJoin:
+		hashJoin.SmallTable = 0
+	case InnerJoin:
+		// We will use right table as small table.
```
Comment (winoros) [inline plan/new_physical_plan_builder.go]: What does this comment mean?

Comment (hanfei1991) [inline plan/new_physical_plan_builder.go]: If right plan's count is smaller than the left one, we will treat right plan as small table.

File: plan/task_profile.go [hunk 2]
```diff
@@ -27,6 +27,7 @@ type taskProfile interface {
 	cost() float64
 	copy() taskProfile
 	plan() PhysicalPlan
+	finishTask(ctx context.Context, allocator *idAllocator) taskProfile
```
Comment (coocood) [inline plan/task_profile.go]: The `finishTask` only make sense in `copTaskProfile`, promote it to `taskProfile` is confusing.

Comment (coocood) [inline plan/task_profile.go]: `finishCopTask` makes more sense.
And add comment on it.

Comment (coocood) [inline plan/task_profile.go]: Move it to a plain function is better, it makes the interface clean.

Comment (hanfei1991) [inline plan/task_profile.go]: fixed, PTAL

**Comment by coocood** [comment]
LGTM

File: plan/task_profile.go [hunk 3]
```diff
@@ -281,9 +309,7 @@ func (p *Union) attach2TaskProfile(profiles ...taskProfile) taskProfile {
 	newTask := &rootTaskProfile{p: np}
 	newChildren := make([]Plan, 0, len(p.children))
 	for _, profile := range profiles {
-		if cop, ok := profile.(*copTaskProfile); ok {
-			profile = cop.finishTask(p.ctx, p.allocator)
-		}
+		profile = finishCopTask(profile, p.ctx, p.allocator)
```
Comment (shenli) [inline plan/task_profile.go]: Why not check task type here?

File: plan/task_profile.go [hunk 4]
```diff
@@ -378,7 +401,7 @@ func (p *PhysicalAggregation) attach2TaskProfile(profiles ...taskProfile) taskPr
 				cop.cnt = cop.cnt * aggFactor
 			}
 		}
-		profile = cop.finishTask(p.ctx, p.allocator)
+		profile = finishCopTask(cop, p.ctx, p.allocator)
```
Comment (shenli) [inline plan/task_profile.go]: Need cop.copy()?

Comment (hanfei1991) [inline plan/task_profile.go]: needn't, finishCopTask will not change cop task.

File: plan/task_profile.go (hunk 3 above)
Comment (hanfei1991) [inline plan/task_profile.go]: task checking is moved to `finishCopTask`

**Comment by shenli** [comment]
LGTM

**Pull request merged**